File: Javascript/Linked List/M-24. Swap Nodes in Pairs.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */
// original: 1 -> 2 -> 3 -> 4
// expected: 2 -> 1 -> 4 -> 3
function swapPairs(head: ListNode | null): ListNode | null {
  if (!head) return head;
  const dummy = new ListNode(-1, head);
  recursive(dummy);
  return dummy.next;
}

const recursive = (node: ListNode | null) => {
  if (!node?.next || !node?.next?.next) return;

  const origNext = node.next;
  const origNextOfNext = node.next.next;
  node.next = origNextOfNext;
  origNext.next = origNextOfNext.next;
  origNextOfNext.next = origNext;
  recursive(origNext);
};

File: Linked List/M-24. Swap Nodes in Pairs/M-24. Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head

        dummy = ListNode(next=head)
        self.recursive(dummy)
        return dummy.next

    def recursive(self, node: Optional[ListNode]):
        if not node or not node.next or not node.next.next:
            return

        next_node = node.next
        next_of_next_node = node.next.next

        node.next = next_of_next_node
        next_node.next = next_of_next_node.next
        next_of_next_node.next = next_node
        self.recursive(next_node)
